src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerAttributeAnalyzer.cs

```diff
@@ -54,7 +54,7 @@ public override void Initialize(AnalysisContext context)
         [SuppressMessage("AnalyzerPerformance", "RS1012:Start action has no registered actions.", Justification = "Method returns an analyzer that is registered by the caller.")]
         protected override DiagnosticAnalyzerSymbolAnalyzer GetDiagnosticAnalyzerSymbolAnalyzer(CompilationStartAnalysisContext compilationContext, INamedTypeSymbol diagnosticAnalyzer, INamedTypeSymbol diagnosticAnalyzerAttribute)
         {
-            var attributeUsageAttribute = WellKnownTypes.AttributeUsageAttribute(compilationContext.Compilation);
+            var attributeUsageAttribute = compilationContext.Compilation.GetTypeByMetadataName(typeof(System.AttributeUsageAttribute).FullName);
```

Comment:
I think we should prioritize on readability and consistency for this as almost every analyzer would need be fetching well known types.
This change seems like the readability has gone down here as each analyzer needs to specify the name of the type. It also introduces the inconsistency that certain well known type names are defined in `WellKnownTypeNames`, but not all. I would instead prefer the following:

1. Consistency: Every queried well known type must have a corresponding constant/static string in `WellKnownTypeNames`. So in this case there will be a public field named `AttributeUsageAttribute` in `WellKnownTypeNames`.
2. Analyzer implementation will be one of the following buckets:
    1. `compilationContext.Compilation.GetTypeByMetadataName(WellKnownTypeNames.AttributeUsageAttribute);`: Used when analyzer wants to fetch a well known type without caching.
   2. `compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.AttributeUsageAttribute, out var wellKnownType);`: Used when analyzer wants to fetch well known type with caching and bail out on null value.
   3. `compilationContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.AttributeUsageAttribute);`: Used when analyzer wants to fetch well known type with caching and continue even on null value.
   4. `var provider = WellKnownTypeProvider.GetOrCreate(compilationContext.Compilation); 
 provider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.AttributeUsageAttribute, out var wellKnownType);`: Used when analyzer needs to query large number of well known types  with caching and/or pass around the provider to different components (such as DFA APIs), but bail out on null return value of type.
   5. `var provider = WellKnownTypeProvider.GetOrCreate(compilationContext.Compilation); 
 provider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.AttributeUsageAttribute);`: Used when analyzer needs to query large number of well known types  with caching and/or pass around the provider to different components (such as DFA APIs), but not bail out on null return value of type.

I am fine if we don't introduce 3. and 5. from above as well.

Reply:
+1 for always using `WellKnownTypeNames`.

Regarding point 2i, I am not sure why you would want to access a type without wanting it be to cached. Indeed, when writing an analyzer you don't really know whether some other will need to access this type. And if you go to the extreme and only look at your analyzer then you might miss some of the benefits of the caching.

Reply:
Sounds reasonable - we could start with just `TryGetOrCreateTypeByMetadataName` methods as our preferred API.

Note that the core `Compilation.GetTypeByMetadataName` API exposed by the compiler is also [caching](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/Compilation.cs,947) well known types, but it is a bounded cache (size 50). IMO, we have not seen any perf traces showing that this cache is not adequate. In general, caching here is just a micro-optimization (whether it is done by the compiler or our well known type provider), which is secondary benefit. The most important aspect here is keeping the analyzer code most readable and have a consistent way to do so.